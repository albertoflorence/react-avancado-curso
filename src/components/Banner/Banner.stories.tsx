import { Story, Meta } from '@storybook/react'
import Banner, { BannerProps } from './Banner'

export default {
  title: 'Banner',
  component: Banner,
  args: {
    image: 'https://source.unsplash.com/user/willianjusten/1042x580',
    title: 'Defy death',
    subtitle: '<p>Play the new <strong>CrashLands</strong> season',
    buttonLabel: 'Buy now',
    buttonLink: '/games/defy-death'
  },
  parameters: {
    layout: 'fullScreen'
  }
} as Meta

export const Default: Story<BannerProps> = args => <Banner {...args} />
