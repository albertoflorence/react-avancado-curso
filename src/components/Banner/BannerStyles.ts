import styled, { css } from 'styled-components'
import media from 'styled-media-query'
import * as Ribbon from 'components/Ribbon/RibbonStyles'

export const Wrapper = styled.div`
  position: relative;
  ${media.greaterThan('medium')`
    box-shadow: 0 0.4rem 0.5rem 0 rgba(0, 0, 0, 0.2);

  `}
  ${media.lessThan('large')`
    ${Ribbon.Wrapper} {
      right: 0;
      &::before {
        display: none;
      }
    }
  `}
`
interface ImageProps {
  src: string
}

export const Image = styled.div<ImageProps>`
  ${({ theme, src }) => css`
    width: 100%;
    height: 230px;
    background-color: ${theme.colors.lightGray};
    background-image: url(${src});
    background-position: center center;
    background-size: cover;

    ${media.greaterThan('medium')`
      height: 580px;
    `}
  `}
`

export const Caption = styled.div`
  ${({ theme }) => css`
    background-color: rgba(0, 0, 0, 0.7);
    padding: ${theme.spacing(3)};
    width: 100%;

    ${media.greaterThan('medium')`
      border-radius: 0 0 ${theme.border.radius} ${theme.border.radius};
      padding: ${theme.spacing(5)};
      position: absolute;
      bottom: 0;
      left: 0;
    `}
  `}
`

export const Title = styled.h2`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.large};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.white};

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.xxlarge};
    `}
  `}
`

export const Subtitle = styled.h3`
  ${({ theme }) => css`
    color: ${theme.colors.white};
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.normal};
    margin-bottom: ${theme.spacing(2)};
    strong {
      color: ${theme.colors.primary};
      font-weight: ${theme.font.bold};
    }
    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.large}
    `}
  `}
`
