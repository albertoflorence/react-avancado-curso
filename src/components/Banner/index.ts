export { default } from './Banner'
export * from './Banner'
export * as BannerStyles from './BannerStyles'
