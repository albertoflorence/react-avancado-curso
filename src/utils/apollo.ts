import { ApolloClient, HttpLink, InMemoryCache, NormalizedCacheObject } from '@apollo/client'
import { setContext } from '@apollo/client/link/context'
import { concatPagination } from '@apollo/client/utilities'
import { Session } from 'next-auth'
import { useMemo } from 'react'

let apolloClient: ApolloClient<NormalizedCacheObject>

function createApolloClient(session?: Session | null) {
  const httpLink = new HttpLink({
    uri: process.env.NEXT_PUBLIC_API_URL + '/graphql'
  })

  const authLink = setContext((_, { headers, session: clientSession }) => {
    const jwt = session?.jwt || clientSession?.jwt || ''

    return {
      headers: {
        ...headers,
        authorization: jwt ? 'Bearer ' + jwt : ''
      }
    }
  })

  return new ApolloClient({
    ssrMode: isServerSide(),
    link: authLink.concat(httpLink),
    cache: apolloCache
  })
}

export const apolloCache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        games: concatPagination(['where', 'sort'])
      }
    }
  }
})

export function initializeApollo(initialState?: NormalizedCacheObject, session?: Session | null) {
  const apolloClientGlobal = apolloClient ?? createApolloClient(session)

  if (initialState) {
    apolloClientGlobal.cache.restore(initialState)
  }

  if (isServerSide()) return apolloClientGlobal

  apolloClient = apolloClient ?? apolloClientGlobal

  return apolloClient
}

export function useApollo(initialState?: NormalizedCacheObject, session?: Session | null) {
  const store = useMemo(() => initializeApollo(initialState, session), [initialState, session])

  return store
}

const isServerSide = () => typeof window === 'undefined'
