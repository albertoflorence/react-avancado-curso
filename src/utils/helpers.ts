export const getImageUrl = (url?: string): string => {
  if (!url) return 'https://source.unsplash.com/random'
  return process.env.NEXT_PUBLIC_API_URL + url
}

export const formatPrice = (price: number): string => {
  return new Intl.NumberFormat('en', { style: 'currency', currency: 'USD' }).format(price)
}

export const formattedPriceToNumber = (str: string): number =>
  Number(str.replace('$', '').replace('.00', ''))
