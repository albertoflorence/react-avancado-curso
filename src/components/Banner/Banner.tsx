import { Button } from 'components'
import * as S from './BannerStyles'

export interface BannerProps {
  image: string
  title: string
  subtitle: string
  buttonLabel: string
  buttonLink: string
}

const Banner = ({ image, title, subtitle, buttonLabel, buttonLink }: BannerProps) => (
  <S.Wrapper>
    <S.Image src={image} role="img" aria-label={title} />
    <S.Caption>
      <S.Title>{title}</S.Title>
      <S.Subtitle dangerouslySetInnerHTML={{ __html: subtitle }} />
      <Button as="a" href={buttonLink} size="large">
        {buttonLabel}
      </Button>
    </S.Caption>
  </S.Wrapper>
)

export default Banner
