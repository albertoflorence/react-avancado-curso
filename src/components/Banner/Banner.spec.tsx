import { render, screen } from 'utils/tests'
import Banner, { BannerProps } from './Banner'

const init = (props: BannerProps) => {
  render(<Banner {...props} />)
  return (name: string, role: string, prop?: string) =>
    screen.getByRole(role, { [prop || 'name']: new RegExp(name, 'i') })
}

const mockBannerProps = (): BannerProps => ({
  image: 'any_image',
  title: 'any_title',
  subtitle: 'any_subtitle',
  buttonLabel: 'any_label',
  buttonLink: 'any_link'
})

describe('<Banner />', () => {
  it('should render correctly', () => {
    const props = mockBannerProps()
    const makeItem = init(props)

    expect(makeItem(props.title, 'heading')).toBeInTheDocument()
    expect(makeItem(props.subtitle, 'heading')).toBeInTheDocument()
    expect(makeItem(props.buttonLabel, 'link')).toBeInTheDocument()
    expect(makeItem(props.image, 'img', 'src')).toBeInTheDocument()
    expect(makeItem(props.buttonLink, 'link', 'href')).toBeInTheDocument()
  })

  it('should render with Ribbon', () => {
    const props = mockBannerProps()
    init({
      ...props,
      ribbon: 'best seller'
    })
    expect(screen.getByText('best seller')).toBeInTheDocument()
  })
})
