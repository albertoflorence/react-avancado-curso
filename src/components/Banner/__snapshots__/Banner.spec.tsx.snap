// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should match the snapshot 1`] = `
.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: none;
  border-radius: 4px;
  padding: 8px;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 400;
  height: 2.5em;
  font-size: 1.125rem;
  padding: 8px 48px;
}

.c5:hover {
  -webkit-filter: brightness(95%);
  filter: brightness(95%);
}

.c0 {
  position: relative;
}

.c1 {
  width: 100%;
  height: 230px;
  background-color: #EAEAEA;
  background-image: url(any_image);
  background-position: center center;
  background-size: cover;
}

.c2 {
  background-color: rgba(0,0,0,0.7);
  padding: 24px;
  width: 100%;
}

.c3 {
  font-size: 1.125rem;
  font-weight: 600;
  color: #FAFAFA;
}

.c4 {
  color: #FAFAFA;
  font-size: 0.875rem;
  font-weight: 400;
  margin-bottom: 16px;
}

.c4 strong {
  color: #F231A5;
  font-weight: 600;
}

@media (min-width:768px) {
  .c0 {
    box-shadow: 0 0.4rem 0.5rem 0 rgba(0,0,0,0.2);
  }
}

@media (min-width:768px) {
  .c1 {
    height: 580px;
  }
}

@media (min-width:768px) {
  .c2 {
    border-radius: 0 0 4px 4px;
    padding: 40px;
    position: absolute;
    bottom: 0;
    left: 0;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 1.75rem;
  }
}

@media (min-width:768px) {
  .c4 {
    font-size: 1.125rem;
  }
}

<div
  class="c0"
>
  <div
    aria-label="any_title"
    class="c1"
    role="img"
    src="any_image"
  />
  <div
    class="c2"
  >
    <h2
      class="c3"
    >
      any_title
    </h2>
    <h3
      class="c4"
    >
      any_subtitle
    </h3>
    <a
      class="c5"
      href="any_link"
    >
      <span>
         
        any_label
         
      </span>
    </a>
  </div>
</div>
`;
